Agents/ForecastingAgent/forecastingagent/agent/forecasting.py

```diff
@@ -0,0 +1,338 @@
+        frequency = self.sparql_client.get_duration_details(input_iris[TS_FREQUENCY])
+        # 4) Data length for training and data scaling
+        data_hist = self.sparql_client.get_duration_details(input_iris[TIME_DURATION])
+
+        # Create forecasting configuration dict
+        cfg = create_forecast_configuration(model=fcmodel, ts_details=ts, ts_frequency=frequency, 
+                                            hist_duration=data_hist, fc_interval=interval)
+        # Add IRI to forecast to config (i.e., IRI which will get ts:hasForecast relationship)
+        cfg['iri_to_forecast'] = input_iris['iri_to_forecast']
+
+        # Create forecast
+        rdb_url, time_format = get_rdb_endpoint(ts)
+        ts_client = TSClient(kg_client=self.sparql_client, rdb_url=rdb_url, 
+                             rdb_user=DB_USER, rdb_password=DB_PASSWORD)
+        fc_ts = forecast(config=cfg, kgClient=self.sparql_client, tsClient=ts_client, 
+                         db_url=rdb_url, time_format=time_format)
+        # Extract times and values from darts TimeSeries
+        times = [str(x) for x in fc_ts.time_index.tz_localize('UTC').strftime(time_format)]
+        values = fc_ts.values().squeeze().tolist()
+        if self.fc_round:
+            # Round data values if set in environment variables
+            values = [round(x, self.fc_round) for x in values]
+
+        # Instantiate new forecast in KG and RDB (if requested or not yet existing)
+        if not self.fc_overwrite or not bool(cfg.get('fc_iri')):
+            # Retrieve potential existing forecast IRI (None if not existing)
+            old_fc_iri = cfg.get('fc_iri')
+            # Initialise forecast in KG
+            fc_iri = self.sparql_client.instantiate_forecast(forecast=fc_ts, config=cfg,
+                                                             create_new=True)
+            # Create new forecast instance and add ts data in RDB
+            ts_client.init_timeseries(dataIRI=fc_iri, times=times, values=values, 
+                                      ts_type=cfg['ts_data_type'], time_format=time_format)
+            if old_fc_iri:
+                # In case new forecast has been created for existing derivation,
+                # disconnect old forecast instance from derivation and connect with new one
+                # NOTE: Derivation Framework only supports one output per rdf type
+                self.sparql_client.reconnect_forecast_with_derivation(old_fc_iri, fc_iri)
+
+            # Add output graph to ensure complete derivation markup
+            # --> this part of the code is only relevant when called via 
+            # 'createSyncDerivationForNewInfo' and its only purpose is to ensure
+            #  that forecast instance is marked up as "belongsTo" the derivation
+            g = Graph()
+            g.add((URIRef(fc_iri), URIRef(RDF_TYPE), URIRef(TS_FORECAST)))   
+            derivation_outputs.addGraph(g)
```

Comment:
For the derivation created when calling `createSyncDerivationForNewInfo`, the standard practice is to add all triples to `derivation_outputs` so that no duplicate information is added to the triple store when the agent handling concurrent update requests. See https://github.com/cambridge-cares/TheWorldAvatar/issues/184 for more information. So I would suggest moving triples in `self.sparql_client.instantiate_forecast` to part of `derivation_outputs`. That said, I suppose the separation here is to support both creating new derivations and updating existing derivations (when not overwriting time series data) - but this goes back to the question of why such a design choice is made.

Reply:
Not all triples from `self.sparql_client.instantiate_forecast` can be added to `derivation_outputs`, as the TimeSeriesClient interacts with the KG directly when writing time series data; however, as overwriting forecasts is now the default, does this "issue" actually still exist?

Reply:
I agree the issue is resolved by itself if "overwriting forecasts" is made as default. But I would recommend documenting this somewhere in the code as the functionality (not overwriting) is still kept.